Add fw_utility test - show, install, update commands

noaOrMlnx opened the pull request:
<!--
Please make sure you've read and understood our contributing guidelines;
https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

Please provide following information to help code review process a bit easier:
-->
### Description of PR
Add test for FW utility feature:
Show, install and update commands.
<!--
- Please include a summary of the change and which issue is fixed.
- Please also include relevant motivation and context. Where should reviewer start? background contaxt?
- List any dependencies that are required for this change.
-->

Summary:
Fixes # (issue)

### Type of change

- [ ] Bug fix
- [ ] Testbed and Framework(new/improvement)
- [X] Test case(new/improvement)

### Approach
#### How did you do it?
N/A
#### How did you verify/test it?
```ini
py.test --inventory ../ansible/inventory --host-pattern <host> --module-path ../ansible/library/ --testbed <host>-<topo> --testbed_file ../ansible/testbed.csv --show-capture=no --capture=no --log-cli-level debug -ra -vvvvv platform/fw_utility/test_fwutil.py --config_file=<path to config file> --binaries_path=<path to binaries> -k <test_case>
```
#### Any platform specific information?
Currently fits to work only on non-modular systems
#### Supported testbed topology if it's a new test case?

### Documentation 
<!--
(If it's a new feature, new test case)
Did you update documentation/Wiki relevant to your implementation?
Link to the wiki page?
-->

noaOrMlnx committed:
Add fw_utility test - show command

nazariig commented on tests/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,117 @@
+
+def fw_status(duthost):
+    stdout = duthost.command("fwutil show status")
+    if stdout['rc'] != 0:
```
@noaOrMlnx please avoid magic numbers

nazariig commented on tests/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,117 @@
+
+
+@pytest.fixture(scope='module')
+def platform_type(duthost):
```
@noaOrMlnx IMHO, we need to define a Jenkins module to gather such a facts

nazariig commented on tests/fw_utility/mellanox_config.yml:
```diff
@@ -0,0 +1,13 @@
+x86_64-mlnx_lssn2700-r0: BIOS,CPLD
```
@noaOrMlnx why do we need two similar config files?

nazariig commented on tests/fw_utility/config/mellanox_config.yml:
```diff
@@ -0,0 +1,13 @@
+x86_64-mlnx_lssn2700-r0: BIOS,CPLD
```
@noaOrMlnx why do we need two similar config files?

nazariig commented on tests/conftest.py:
```diff
@@ -58,6 +58,11 @@ def pytest_addoption(parser):
 
+    # fw_utility options
+    parser.addoption("--config_file", action="store", default=None, help="name of configuration file (per each vendor)")
+    parser.addoption("--binaries_path", action="store", default=None, help="path to binaries files")
```
@noaOrMlnx nice catch

nazariig commented on tests/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,117 @@
+COMPONENTS = {'BIOS', 'CPLD'}
+BASE_DIR = os.path.dirname(os.path.realpath(__file__))
+BINARIES_DIR = os.path.join(BASE_DIR, 'binaries')
+TMP_DIR = os.path.basename('tmp')
```
@noaOrMlnx please use absolute paths unless you have a reasons not to do it

noaOrMlnx committed:
Fix review comments
- Magic numbers
- Using module facts instead separate function
- Remove unnecassary files

Signed-off-by: Noa Or <noaor@mellanox.com>

nazariig commented on tests/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,86 @@
+logger = logging.getLogger(__name__)
+
+FW_UTIL_DATA = {}
+COMPONENTS = {'BIOS', 'CPLD'}
```
@noaOrMlnx why do we need this? Cause we already have `tests/fw_utility/config/mellanox_config.yml`

noaOrMlnx committed:
Fix review comments

noaOrMlnx committed:
Add install command to fwutil test

noaOrMlnx committed:
Update test_fwutil.py

nazariig commented on tests/platform/fw_utility/config/mellanox_config.yml:
```diff
@@ -0,0 +1,13 @@
+x86_64-mlnx_lssn2700-r0: BIOS,CPLD
```
@noaOrMlnx The fwutil supports both modular and non modular chassis platforms. Do we want to make a corresponding infrastructure to be available in py.test?

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+
+def test_install_negative(request, testbed_devices, get_fw_path):
+    dut = testbed_devices["dut"]
+    pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig replied:
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+
+
+@pytest.fixture(scope='function')
+def install_local(request, testbed_devices, testbed, get_fw_path, components_list):
```
@noaOrMlnx can we slightly improve the entire infrastructure and move all that fixtures to the conftest file just to avoid test environment pollution? What do you think?

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    time.sleep(20)
+
+
+def check_bios_version(version_to_install, comp_data):
```
@noaOrMlnx for auxiliary stuff like this i would recommend to create a separate class / module.

noaOrMlnx replied:
Done

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    """
+    stdout = dut.command("fwutil show status")
+    if stdout['rc'] != SUCCESS_CODE:
+        err = stdout['stderr']
```
@noaOrMlnx variables naming looks like invalid to me.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+import os
+import pdb
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    assert check_pmon_daemon_status(dut), "Not all pmon daemons running."
+
+    if dut.facts["asic_type"] in ["mellanox"]:
+        pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    """
+    performs 30 sec power cycle off to finish cpld installation.
+    """
+    pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    psu_control = request.getfixturevalue("psu_ctrl")
+    if psu_control is None:
+        pytest.fail("No PSU controller for %s, skip rest of the testing in this case" % dut.hostname)
+    pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    # complete fw update - cold reboot if BIOS, power cycle with 30 sec timeout if CPLD
+    globals()[component.lower() + '_update'](request, dut)
+    # check output of show command
+    pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+
+    # recover previous fw only if previous was the latest:
+    if get_fw_path['is_latest_installed'] == True:
+        pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    """
+    return psu_controller
+    """
+    pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+            pytest.fail("Installation Failed. Aborting!")
+
+    yield
+    pdb.set_trace()
```
@noaOrMlnx we don't need debug stuff in production.

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,501 @@
+    execute wrong commands for installation and check the right errors appear.
+    """
+    cmd = 'fwutil install chassis component {} fw -y {}'.format(comp_name, comp_path)
+    stdout = dut.command(cmd, module_ignore_errors=True)
```
@noaOrMlnx Suggest to replace:`stdout ` -> `result`

noaOrMlnx committed:
Fix review comments
- Change table parsing
- Create dedicated conftest for the test
- Create file contains help function

Signed-off-by: Noa Or <noaor@mellanox.com>

noaOrMlnx committed:
Update fwutil_helper.py

noaOrMlnx committed:
Update fwutil_helper.py

noaOrMlnx committed:
Add update (positive & negative) test cases.

Signed-off-by: Noa Or <noaor@mellanox.com>

noaOrMlnx committed:
Merge branch 'noaor_fwutility' of https://github.com/noaOrMlnx/sonic-mgmt into noaor_fwutility

noaOrMlnx committed:
Update conftest.py

noaOrMlnx committed:
Update test_fwutil.py

noaOrMlnx committed:
Update conftest.py with extra option

noaOrMlnx committed:
Update conftest.py with extra option

noaOrMlnx committed:
Update conftest.py with extra option

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,256 @@
+        assert False, "Unable to get the number of PSUs using command '%s'" % cmd_num_psu
+
+    logging.info("Create PSU controller for testing")
+    psu_control = request.getfixturevalue("psu_ctrl")
```
@noaOrMlnx why do we need a whole Fixture here? What is the benefits?

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,570 @@
+logger = logging.getLogger(__name__)
+
+
+def fw_status(dut):
```
@noaOrMlnx why not to move to fwutil_helper.py?

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,256 @@
+    # for now just checks if ends without errors - without version check
+    return SUCCESS_CODE
+
+    # counts = {}
```
@noaOrMlnx why do we have this in production code?

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,256 @@
+
+    # logging.info("Check the uptime to verify whether reboot was performed")
+    # dut_uptime = datetime.strptime(dut.command("uptime -s")["stdout"], "%Y-%m-%d %H:%M:%S")
+    # assert float(dut_uptime.strftime("%s")) - float(dut_datetime.strftime("%s")) > 10, "Device did not reboot"
```
@noaOrMlnx why do we have this in production code?

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,256 @@
+from check_daemon_status import check_pmon_daemon_status
+
+import logging
+logger = logging.getLogger(__name__)
```
@noaOrMlnx do we need to specify a logger channel (stdout/file or both)?

nazariig commented on tests/platform/fw_utility/conftest.py:
```diff
@@ -0,0 +1,52 @@
+    """
+    backup the original platform_components.json file
+    """
+    dut = testbed_devices["dut"]
```
@noaOrMlnx can we use a common style for all dict access (e.g, "" or '')?

nazariig commented on tests/platform/fw_utility/conftest.py:
```diff
@@ -0,0 +1,52 @@
+import os
+import pytest
+from fwutil_helper import *
```
@noaOrMlnx why do you need to import all?

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,256 @@
+import time
+import sys
+from common import reboot
+from psu_controller import psu_controller
```
@noaOrMlnx why do you need psu_controller here?

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,570 @@
+    result = dut.command("fwutil show status")
+    if result['rc'] != SUCCESS_CODE:
+        pytest.fail("Could not execute command 'fwutil show status'")
+        raise result['stderr']
```
@noaOrMlnx why do we need raise? Won't the pytest.fail produce the exception?

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,570 @@
+    Purpose of the test is checking all required components appear in
+    'fwutil show status' command output according the given config file.
+    :param dut: dut
+    :param fw_status: fixture for getting status.
```
@noaOrMlnx please fix the description

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,570 @@
+    return result['stdout']
+
+
+def get_output_data(dut):
```
@noaOrMlnx why not to move to fwutil_helper.py?

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,570 @@
+    if not comp_data['version']:
+        pytest.fail("Installation didn't work. Aborting!")
+
+    return globals()['check_' + component.lower() + '_version'](version_to_install, comp_data)
```
@noaOrMlnx looks like a dirty hack: + for your creativity but - for the design

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,256 @@
+
+def check_cpld_version(version_to_install, comp_data):
+
+    # for now just checks if ends without errors - without version check
```
@noaOrMlnx what exactly we check here?

noaOrMlnx committed:
Fix review comments
- Add objects to FW components
- Clean test file to contain only test cases
- Conftest contains fixtures

Signed-off-by: Noa Or <noaor@mellanox.com>

noaOrMlnx committed:
Clean files from debug related lines

Signed-off-by: Noa Or <noaor@mellanox.com>

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,617 @@
+                r = r.split('_')[0]
+                revisions.append(int(r[0:2]))
+            else:
+                r = r.split('.')[0]
```
@noaOrMlnx this looks like a copy-paste from BIOS

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,617 @@
+    def get_component_name(self):
+        pass
+
+    def process_versions(self, dut, components_list, location):
```
@noaOrMlnx why do we need `components_list` here?

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,617 @@
+class FwComponent():
+
+    def get_version(self, dut, files_path, fw_data):
+        pass
```
@noaOrMlnx should be `raise NotImplemented()`

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,617 @@
+INVALID_COMPONENT_SCHEMA_LOG = '.*Error: Failed to parse "platform_components.json": invalid component schema*.'
+
+
+class FwComponent():
```
@noaOrMlnx It's a good practise to always inherit from `object`

nazariig commented on tests/platform/fw_utility/test_fwutil.py:
```diff
@@ -0,0 +1,114 @@
+    NOTICE: The next image should be an image with fwutil feature
+    """
+    dut = testbed_devices['dut']
+    update_from_current_img(request, dut, get_fw_path, components_list, component_object)
```
@noaOrMlnx If am not mistaken, there is no guarantee that `component_object` and `get_fw_path` results will be relevant, at least always. What if the system will randomly pick up BIOS and CPLD FW path? The variety of combinations will grow with adding a new platform components. Please correct me if i am wrong.

nazariig commented on tests/platform/fw_utility/fwutil_helper.py:
```diff
@@ -0,0 +1,617 @@
+
+        versions = {
+            'latest_version': latest_ver,
+            'latest_path': latest_fw_path,
```
@noaOrMlnx Looks like we have mismatch here since we are using current FW version in system + latest FW path

noaOrMlnx committed:
Fix review comments
